### DISCUSSION TITLE
Generative adversarial network

### DISCUSSION THREAD

**Discussion by wangjiawen2013** [tracel-ai/burn#2593] [Q&A]
Hi,
I am working on a GAN recently like this: 
![image](https://github.com/user-attachments/assets/6525d472-4c63-4313-92f1-56850aa4bb5a)

In each epoch, I must train the discriminator for many times and then the generator for one time. I find it's hard to implement GAN using `learnerbuilder` similar to the code in burn example:
![image](https://github.com/user-attachments/assets/e02ea38d-bc0b-4821-b9b0-96fb8358b4cc)

Does `learnerbuilder` applies only to  simple network ? Do I need to write customized loss function and customized training loop ?

**Comment by laggui** [comment]
Training GANs requires training a generator and a discriminator, with one optimizer for each.

This is a bit more complex than the typical training loop implemented by the learner so you would have to roll with your own training loop 🙂

**Reply by wangjiawen2013** [reply]
I am re-implementing pytorch wgan with burn recently days, this is the github repository: https://github.com/wangjiawen2013/wgan

I want to merge it into burn's github examples after finishing it. But some problems are still needed to addressed, so may I ask your help here? I hope that you can have a look at it and find the failure reasons.

I have run pytorch wgan sucessfully, this is the output:
![170000](https://github.com/user-attachments/assets/03e5842e-7c41-4135-a51e-2592c6493261)

But for burn wgan, it is just meanless and repeated patterns:
![image900](https://github.com/user-attachments/assets/2fa2dad8-322c-4b88-b690-16f5f067b8c5)

There must be something wrong with my code. I have tested a lot of times and find that most things are as expected. I have strong doubts that the `loss` or the `optimizer.step()` don't work (src/training.rs line 121-130 and line 135-142). It is expected that the loss will be **_updated fastly_** when running optimizer.step() many times. In pytorch wgan, it worked as expected. But in burn wgan, the loss (D loss: discriminator loss, G loss: generator loss) **_updated slowly_** when running optimizer.step(), no matter it is `let loss_d = -discriminator.forward(batch.images).mean() + discriminator.forward(fake_images).mean();` or `let loss_d = discriminator.forward(batch.images).mean() - discriminator.forward(fake_images).mean();` !!!  And the D loss and G loss differed a lot from that of pytorch, **_approximate several hundreds times lower than pytorch_**. Here is the training process:
```text
[Epoch 1/3] [Batch 0/938] [D loss: -0.09546152] [G loss: -0.010683101]
[Epoch 1/3] [Batch 10/938] [D loss: -0.09064701] [G loss: -0.010889139]
[Epoch 1/3] [Batch 20/938] [D loss: -0.09109107] [G loss: -0.011392368]
[Epoch 1/3] [Batch 30/938] [D loss: -0.0905629] [G loss: -0.012158312]
[Epoch 1/3] [Batch 40/938] [D loss: -0.08930381] [G loss: -0.013122454]
[Epoch 1/3] [Batch 50/938] [D loss: -0.088740215] [G loss: -0.014366182]
[Epoch 1/3] [Batch 60/938] [D loss: -0.0866037] [G loss: -0.016056392]
[Epoch 1/3] [Batch 70/938] [D loss: -0.08377892] [G loss: -0.018405747]
[Epoch 1/3] [Batch 80/938] [D loss: -0.08068731] [G loss: -0.021662386]
[Epoch 1/3] [Batch 90/938] [D loss: -0.07828338] [G loss: -0.026396101]
[Epoch 1/3] [Batch 100/938] [D loss: -0.067133166] [G loss: -0.03337042]
[Epoch 1/3] [Batch 110/938] [D loss: -0.057932906] [G loss: -0.04426057]
[Epoch 1/3] [Batch 120/938] [D loss: -0.04228705] [G loss: -0.059334144]
[Epoch 1/3] [Batch 130/938] [D loss: -0.021414958] [G loss: -0.08037117]
[Epoch 1/3] [Batch 140/938] [D loss: 0.0035974532] [G loss: -0.10367289]
[Epoch 1/3] [Batch 150/938] [D loss: 0.021838397] [G loss: -0.1235547]
[Epoch 1/3] [Batch 160/938] [D loss: 0.034707725] [G loss: -0.13705881]
[Epoch 1/3] [Batch 170/938] [D loss: 0.046433717] [G loss: -0.14530241]
[Epoch 1/3] [Batch 180/938] [D loss: 0.048114464] [G loss: -0.15006739]
[Epoch 1/3] [Batch 190/938] [D loss: 0.05085115] [G loss: -0.1528728]
[Epoch 1/3] [Batch 200/938] [D loss: 0.051142164] [G loss: -0.15466571]
```
I know it is a time-consuming labour to read through these code, while I hope that you could give some suggestions to help it works in burn.

**Comment by wangjiawen2013** [comment]
Here is the training process of pytorch wgan, it worked as expected:

![image](https://github.com/user-attachments/assets/802b9d3e-0aa9-42c8-bc67-bfde964e0fb5)
